### PR TITLE
Add logic to get results from the search bar action menu clicks

### PR DESCRIPTION
**Pull request by Akrion**
Added `actionWrapper` which when passed to the `SearchBar/TagsQuerySearchBar` allows for a feedback of the `ActionMenu` actions via the `after` and `onError` aspects.

### CODE REVIEW

**Comment by decrapifier** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Your PR has lint errors. Please fix these and commit them.
      
<details>
  <summary>ActionsMenu.js</summary>
  <p><a href="https://github.com/smartprocure/contexture-react/blob/feature/ActionBarFeedbackFn/src/exampleTypes/TagsQuery/ActionsMenu.js" target="_blank">src/exampleTypes/TagsQuery/ActionsMenu.js</a></p>
  
  <strong>'F' is defined but never used.</strong>
  <code>Line 3: import F from 'futil-js'</code>
</details>
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>The changelog has not been updated. Please update the changelog.</td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>The version was not updated. Please update the version.</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>The README has not been updated. Please update the README.</td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  Could not run `duti:fix` command successfully
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>Could not find any browser results.</td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against c263187ed3d25981c57178c7b6c2fdc0a7ce1d73
</p>

**Comment by decrapifier** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>The changelog has not been updated. Please update the changelog.</td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>The version was not updated. Please update the version.</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>The README has not been updated. Please update the README.</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>Could not find any browser results.</td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 506314ae8d476434f6e6fedeb098fbea9b49a452
</p>

**Comment by stellarhoof** [comment]
Why not just display an error under the action?

**Comment by Akrion** [comment]
Usually the common way to do this is the actual action menu closes and you get feedback via the `toast`. I would think that is the expected these days UI feedback. I do not think that the error should be on the action menu ... it is kind of weird. It is not a form per-se etc.

**Comment by stellarhoof** [comment]
> Usually the common way to do this is the actual action menu closes and you get feedback via the toast

Or that. We've got toasts, errors under the component or even a commandbutton replica with a console error. They all seem better to me than this. I mean, there are so many actions performed everywhere that don't have callbacks, nor they should because it'd increase the complexity of the component so much. Just think resulttable uff

**Comment by Akrion** [comment]
Well that was the intention ... this now allows for the main UI to show `toasts` based on the argument passed in the function `onActionMenuClick`. We would not want to couple and have an opinion about toast when it comes to `contexture-react` wold be my guess right?

Not set on this but we need something to give us feedback that action menu was successful or not and use the consumer of contexture-react to provide the user with the feedback.

**Comment by stellarhoof** [comment]
Yeah toasts are opinionated for sure, which is why my 2c is to just show the error under the action.

**Comment by Akrion** [comment]
Yeah not a fan of that ... from UI stand point just feels weird. To me weirder than sending result back via the argument. But if this is the overall opinion I will agree etc.

**Comment by stellarhoof** [comment]
Ya let's just get more eyez on this

File: src/exampleTypes/TagsQuery/ActionsMenu.js [hunk 1]
```diff
@@ -1,57 +1,57 @@
...
+    <Flex
+      style={{ minWidth: 240, padding: 10 }}
+      className="tags-query-actions-menu"
+      column
+      justifyContent="stretch"
+      alignItems="stretch"
+    >
+      {!!_.get('tags.length', node) && (
+        <>
+          <Button onClick={() => wrapper('copy', () => copyTags(node))}>
```
Comment (daedalus28) [inline src/exampleTypes/TagsQuery/ActionsMenu.js]: I think we can simplify this if we instead just take a wrapper and assume that the return values are success messages and errors are the error messages. On the user side, we'd have one of these that just logs that with toastify:

https://runkit.com/daedalus28/toasty-futil-aspect

Comment (Akrion) [reply src/exampleTypes/TagsQuery/ActionsMenu.js]: Changes done.

File: src/stories/imdb/stories/searchLayout.js [hunk 2]
```diff
@@ -258,6 +260,7 @@ let GreyVestSearchBarStory = theme => (
               path={['root', 'bar']}
               resultsPath={['root', 'results']}
               autoFocus
+              actionWrapper={aspectWrapper}
```
Comment (Akrion) [inline src/stories/imdb/stories/searchLayout.js]: Let this here as a test/demo of the idea of `aspectWrapper`

File: src/utils/futil.js [hunk 3]
```diff
@@ -38,3 +38,8 @@ export let expandObject = _.curry((transform, obj) => ({
 export let expandObjectBy = _.curry((key, fn, obj) =>
   expandObject(getWith(fn, key))(obj)
 )
+
+export let aspectWrapper = F.aspect({
```
Comment (Akrion) [inline src/utils/futil.js]: Kept this here for demo purposes.

**Comment by Akrion** [comment]
Nothing is set in stone for this and I am open for suggestions. If we are saying that we would not need any feedback from the `action button` and we will handle all that is needed at `contexture-react` level by disabling buttons etc I am fine with that too. We just need to choose a direction/pattern and go with it.

File: src/utils/futil.js [hunk 4]
```diff
@@ -38,3 +38,9 @@ export let expandObject = _.curry((transform, obj) => ({
 export let expandObjectBy = _.curry((key, fn, obj) =>
   expandObject(getWith(fn, key))(obj)
 )
+
+export let aspectWrapper = open => F.aspect({
```
Comment (daedalus28) [inline src/utils/futil.js]: I think we can simplify this even further if we don't make the wrapper aware of `open` - I don't think end users are going to want to control the open-ness from the aspect

Comment (daedalus28) [reply src/utils/futil.js]: but testing this idea here was really smart 👏

Comment (Akrion) [reply src/utils/futil.js]: Changes done.

**Comment by stellarhoof** [comment]
@Akrion You can do `_.has('clipboard', navigator)` or something and disable the button accordingly. Haven't tested in Safari since I don't have a Mac. The second action is a `tree.mutate`. To my earlier point on the blurry lines of error handling: we do contexture stuff like this all over the place so I see no point in wrapping this one.

**Comment by Akrion** [comment]
@stellarhoof I do see your point however I also do believe the user should get a feedback saying that his tags were (or were not) copied into the clipboard. It is a common pattern and you kind of intuitively expect it when you do `copy` to clipboard. Otherwise we really are leaving the user guessing did it work or not and having him "testing" if it did via pasting somewhere. 

So I am not aware of many other places where such a feedback is necessary but if there are we should discuss them as well. It might be that we should also wrap some of them as well. 

Now I am not saying this "wrapper" approach is the best but it seems to be the most out of the box we could do since we already have the AOP support in `Futil` with the `aspect`. Maybe we have a better way to wrap or expose the action result etc ...

**Comment by stellarhoof** [comment]
So do you want to 1) Always have some visual feedback on whether copying something was successful or 2) Show a toast when there's an error? If we want 1) we'd be better off with passing a custom copy function everywhere that displays a toast on success. Another advantage of this is that we can use a polyfill to make sure we always succeed.

File: src/exampleTypes/TagsQuery/ActionsMenu.js [hunk 5]
```diff
@@ -7,7 +7,13 @@ import { withTheme } from '../../utils/theme'
 import { Flex } from '../../greyVest'
 import { copyTags, tagTerm } from './utils'
 
-let ActionsMenu = ({ node, tree, open, theme: { Button, Checkbox } }) => (
+let ActionsMenu = ({
+  node,
+  tree,
+  open,
+  theme: { Button, Checkbox },
+  actionWrapper = _.noop,
```
Comment (daedalus28) [inline src/exampleTypes/TagsQuery/ActionsMenu.js]: I think should default to `_.identity`. If you omit `actionWrapper`, `_.noop` returns undefined which would prevent the onClicks from doing anything.

File: package.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "contexture-react",
-  "version": "2.7.2",
+  "version": "2.7.3",
```
Comment (daedalus28) [inline package.json]: sorry for the churn, just need to update the version

**Comment by daedalus28** [comment]
@stellarhoof the point of this PR is to give visible feedback for these buttons, so I think this is the right approach. We're actually less concerned about errors on copy, and more concerned that the success case currently gives no visual feedback so users have no idea if it worked.

**Comment by stellarhoof** [comment]
@daedalus28 See my previous comment https://github.com/smartprocure/contexture-react/pull/354#issuecomment-552020185

**Comment by Akrion** [comment]
Fixed

**Pull request merged**